### PR TITLE
repair the yolov8n network

### PR DESCRIPTION
**Pull request by yupeng-2003**
repair the yolov8n network

### CODE REVIEW

**Comment by lindsayshuo** [comment]
> repair the yolov8n network
Just modify the code in the following areas, and the code in other areas is correct:
// output0
    int ch = std::max(64,std::min(kNumClass,100));
    nvinfer1::IElementWiseLayer* conv22_cv2_0_0 = convBnSiLU(network, weightMap, *conv15->getOutput(0), 64, 3, 1, 1, "model.22.cv2.0.0");
    nvinfer1::IElementWiseLayer* conv22_cv2_0_1 = convBnSiLU(network, weightMap, *conv22_cv2_0_0->getOutput(0), 64, 3, 1, 1, "model.22.cv2.0.1");
    nvinfer1::IConvolutionLayer* conv22_cv2_0_2 = network->addConvolutionNd(*conv22_cv2_0_1->getOutput(0), 64, nvinfer1::DimsHW{1,1}, weightMap["model.22.cv2.0.2.weight"], weightMap["model.22.cv2.0.2.bias"]);
    conv22_cv2_0_2->setStrideNd(nvinfer1::DimsHW{1, 1});
    conv22_cv2_0_2->setPaddingNd(nvinfer1::DimsHW{0, 0});

    nvinfer1::IElementWiseLayer* conv22_cv3_0_0 = convBnSiLU(network, weightMap, *conv15->getOutput(0), ch, 3, 1, 1, "model.22.cv3.0.0");
    nvinfer1::IElementWiseLayer* conv22_cv3_0_1 = convBnSiLU(network, weightMap, *conv22_cv3_0_0->getOutput(0), ch, 3, 1, 1, "model.22.cv3.0.1");
    nvinfer1::IConvolutionLayer* conv22_cv3_0_2 = network->addConvolutionNd(*conv22_cv3_0_1->getOutput(0), kNumClass, nvinfer1::DimsHW{1,1}, weightMap["model.22.cv3.0.2.weight"], weightMap["model.22.cv3.0.2.bias"]);
    conv22_cv3_0_2->setStride(nvinfer1::DimsHW{1, 1});
    conv22_cv3_0_2->setPadding(nvinfer1::DimsHW{0, 0});
    nvinfer1::ITensor* inputTensor22_0[] = {conv22_cv2_0_2->getOutput(0), conv22_cv3_0_2->getOutput(0)};
    nvinfer1::IConcatenationLayer* cat22_0 = network->addConcatenation(inputTensor22_0, 2);

    // output1
    nvinfer1::IElementWiseLayer* conv22_cv2_1_0 = convBnSiLU(network, weightMap, *conv18->getOutput(0), 64, 3, 1, 1, "model.22.cv2.1.0");
    nvinfer1::IElementWiseLayer* conv22_cv2_1_1 = convBnSiLU(network, weightMap, *conv22_cv2_1_0->getOutput(0), 64, 3, 1, 1, "model.22.cv2.1.1");
    nvinfer1::IConvolutionLayer* conv22_cv2_1_2 = network->addConvolutionNd(*conv22_cv2_1_1->getOutput(0), 64, nvinfer1::DimsHW{1, 1}, weightMap["model.22.cv2.1.2.weight"], weightMap["model.22.cv2.1.2.bias"]);
    conv22_cv2_1_2->setStrideNd(nvinfer1::DimsHW{1,1});
    conv22_cv2_1_2->setPaddingNd(nvinfer1::DimsHW{0,0});

    nvinfer1::IElementWiseLayer* conv22_cv3_1_0 = convBnSiLU(network, weightMap, *conv18->getOutput(0), ch, 3, 1, 1, "model.22.cv3.1.0");
    nvinfer1::IElementWiseLayer* conv22_cv3_1_1 = convBnSiLU(network, weightMap, *conv22_cv3_1_0->getOutput(0), ch, 3, 1, 1, "model.22.cv3.1.1");
    nvinfer1::IConvolutionLayer* conv22_cv3_1_2 = network->addConvolutionNd(*conv22_cv3_1_1->getOutput(0), kNumClass, nvinfer1::DimsHW{1, 1}, weightMap["model.22.cv3.1.2.weight"], weightMap["model.22.cv3.1.2.bias"]);
    conv22_cv3_1_2->setStrideNd(nvinfer1::DimsHW{1,1});
    conv22_cv3_1_2->setPaddingNd(nvinfer1::DimsHW{0,0});

    nvinfer1::ITensor* inputTensor22_1[] = {conv22_cv2_1_2->getOutput(0), conv22_cv3_1_2->getOutput(0)};
    nvinfer1::IConcatenationLayer* cat22_1 = network->addConcatenation(inputTensor22_1, 2);

    // output2
    nvinfer1::IElementWiseLayer* conv22_cv2_2_0 = convBnSiLU(network, weightMap, *conv21->getOutput(0), 64, 3, 1, 1, "model.22.cv2.2.0");
    nvinfer1::IElementWiseLayer* conv22_cv2_2_1 = convBnSiLU(network, weightMap, *conv22_cv2_2_0->getOutput(0), 64, 3, 1, 1, "model.22.cv2.2.1");
    nvinfer1::IConvolutionLayer* conv22_cv2_2_2 = network->addConvolution(*conv22_cv2_2_1->getOutput(0), 64, nvinfer1::DimsHW{1,1}, weightMap["model.22.cv2.2.2.weight"], weightMap["model.22.cv2.2.2.bias"]);

    nvinfer1::IElementWiseLayer* conv22_cv3_2_0 = convBnSiLU(network, weightMap, *conv21->getOutput(0), ch, 3, 1, 1, "model.22.cv3.2.0");
    nvinfer1::IElementWiseLayer* conv22_cv3_2_1 = convBnSiLU(network, weightMap, *conv22_cv3_2_0->getOutput(0), ch, 3, 1, 1, "model.22.cv3.2.1");
    nvinfer1::IConvolutionLayer* conv22_cv3_2_2 = network->addConvolution(*conv22_cv3_2_1->getOutput(0), kNumClass, nvinfer1::DimsHW{1,1}, weightMap["model.22.cv3.2.2.weight"], weightMap["model.22.cv3.2.2.bias"]);

    nvinfer1::ITensor* inputTensor22_2[] = {conv22_cv2_2_2->getOutput(0), conv22_cv3_2_2->getOutput(0)};
    nvinfer1::IConcatenationLayer* cat22_2 = network->addConcatenation(inputTensor22_2, 2);

File: yolov8/src/model.cpp [hunk 1]
```diff
@@ -41,7 +42,7 @@ nvinfer1::IHostMemory* buildEngineYolov8n(nvinfer1::IBuilder* builder,
     nvinfer1::ITensor* inputTensor11[] = {upsample10->getOutput(0), conv6->getOutput(0)};
     nvinfer1::IConcatenationLayer* cat11 = network->addConcatenation(inputTensor11, 2);
 
-    nvinfer1::IElementWiseLayer* conv12 = C2F(network, weightMap, *cat11->getOutput(0), 128, 128, 1, false, 0.5, "model.12");
+    nvinfer1::IElementWiseLayer* conv12 = C2F(network, weightMap, *cat11->getOutput(0), 384,128, 1, false, 0.5, "model.12");
```
Comment (wang-xinyu) [inline yolov8/src/model.cpp:45]: 类似这里的layer修改，是什么原因呢？  原来的值， @lindsayshuo 都是验证过的

Comment (yupeng-2003) [reply yolov8/src/model.cpp:45]: 这是根据网络结构各层的输入输出通道数得到的，以v8n为例，12层输入通道数是384，输出通道数是128

不过在后面并没有用到输入通道数，不会影响结果

Comment (wang-xinyu) [reply yolov8/src/model.cpp:45]: 好的，那是不是只需要加这个max的逻辑能够修复问题了？`int ch = std::max(64,std::min(kNumClass,100));`

Comment (yupeng-2003) [reply yolov8/src/model.cpp:45]: 是的

**Comment by wang-xinyu** [comment]
@yupeng-2003 目前有另外一个PR（https://github.com/wang-xinyu/tensorrtx/pull/1381 ）正在对yolov8构建网络这块的代码做重构，为了避免冲突，需要等那个PR合入之后，你这里再适配一下，再等等哈。    

另外这个max逻辑是只需要加在yolov8n吗？

**Comment by yupeng-2003** [comment]
> @yupeng-2003 目前有另外一个PR（#1381 ）正在对yolov8构建网络这块的代码做重构，为了避免冲突，需要等那个PR合入之后，你这里再适配一下，再等等哈。
> 
> 另外这个max逻辑是只需要加在yolov8n吗？

好的。是的，其他的也是这个逻辑，但其他的网络大，取max就是固定的值。

**Comment by yupeng-2003** [comment]
> @yupeng-2003 #1381 已经合入master了，帮忙看看你这个修改还需要吗？ 好像他那边也有max的逻辑

我认为对于v8n还是会有问题。 [#1831](https://github.com/wang-xinyu/tensorrtx/pull/1381) 中的max只是防止超出最大通道数max_channels。对于v8n，得到的output_channel是固定的80，这对于80类的coco数据集没问题。但按照 ch = std::max(64,std::min(kNumClass,100))，如果是2类的visdrone等小于64类的数据集，output_channel就应该是64；如果类别数kNumClass在64-80间，output_channel就应该是kNumClass。所以对于v8n还是应该加入max。

**Comment by wang-xinyu** [comment]
@yupeng-2003 那能否在最新的代码基础上再提个PR~~？  

cc: @Rex-LK

**Comment by Rex-LK** [comment]
> @yupeng-2003 那能否在最新的代码基础上再提个PR~~？
> 
> cc: @Rex-LK
好的，我试试！

**Comment by Rex-LK** [comment]
> > @yupeng-2003 那能否在最新的代码基础上再提个PR~~？
> > cc: @Rex-LK
> > 好的，我试试！



> > @yupeng-2003 #1381 已经合入master了，帮忙看看你这个修改还需要吗？ 好像他那边也有max的逻辑
> 
> 我认为对于v8n还是会有问题。 [#1831](https://github.com/wang-xinyu/tensorrtx/pull/1381) 中的max只是防止超出最大通道数max_channels。对于v8n，得到的output_channel是固定的80，这对于80类的coco数据集没问题。但按照 ch = std::max(64,std::min(kNumClass,100))，如果是2类的visdrone等小于64类的数据集，output_channel就应该是64；如果类别数kNumClass在64-80间，output_channel就应该是kNumClass。所以对于v8n还是应该加入max。

我在[#1387](https://github.com/wang-xinyu/tensorrtx/pull/1387)上提交了关于yolo8n的ch的计算，看是不是你预期的呢? @yupeng-2003 @wang-xinyu

**Comment by yupeng-2003** [comment]
> > > @yupeng-2003 那能否在最新的代码基础上再提个PR~~？
> > > cc: @Rex-LK
> > > 好的，我试试！
> 
> > > @yupeng-2003 #1381 已经合入master了，帮忙看看你这个修改还需要吗？ 好像他那边也有max的逻辑
> > 
> > 
> > 我认为对于v8n还是会有问题。 [#1831](https://github.com/wang-xinyu/tensorrtx/pull/1381) 中的max只是防止超出最大通道数max_channels。对于v8n，得到的output_channel是固定的80，这对于80类的coco数据集没问题。但按照 ch = std::max(64,std::min(kNumClass,100))，如果是2类的visdrone等小于64类的数据集，output_channel就应该是64；如果类别数kNumClass在64-80间，output_channel就应该是kNumClass。所以对于v8n还是应该加入max。
> 
> 我在[#1387](https://github.com/wang-xinyu/tensorrtx/pull/1387)上提交了关于yolo8n的ch的计算，看是不是你预期的呢? @yupeng-2003 @wang-xinyu

应该没问题了

**Pull request closed without merging**